### PR TITLE
Adding the new endpoint for keys

### PR DESCRIPTION
**Pull request by cody-constine-ttd**

### CODE REVIEW

File: src/main/java/com/uid2/core/service/IKeysetMetadataProvider.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.uid2.core.service;
+
+import com.uid2.core.util.OperatorInfo;
+
+public interface IKeysetMetadataProvider {
```
Comment (jon8787) [inline src/main/java/com/uid2/core/service/IKeysetMetadataProvider.java:5]: I can't see IKeysetMetadataProvider being consumed anywhere, only implemented. Can we remove it?

File: src/main/java/com/uid2/core/vertx/CoreVerticle.java [hunk 2]
```diff
@@ -264,6 +270,28 @@ private void handleKeyAclRefresh(RoutingContext rc) {
...
     }
 
+    private void handleKeysetRefresh(RoutingContext rc) {
+        try {
+            OperatorInfo info = OperatorInfo.getOperatorInfo(rc);
+            rc.response().putHeader(HttpHeaders.CONTENT_TYPE, "application/json")
+                    .end(keysetMetadataProvider.getMetadata(info));
+        } catch (Exception e) {
+            logger.warn("exception in handleKeysetRefresh: " + e.getMessage(), e);
+            Error("error", 500, rc, "error processing key refresh");
```
Comment (jon8787) [inline src/main/java/com/uid2/core/vertx/CoreVerticle.java:280]: should Error() do the logging instead? When would we want Error() without a log?

File: src/main/java/com/uid2/core/vertx/CoreVerticle.java [hunk 3]
```diff
@@ -126,6 +130,8 @@ private Router createRoutesSetup() {
                 .handler(auth.handle(this::handleAttestAsync, Role.OPERATOR));
         router.get("/key/refresh").handler(auth.handle(attestationMiddleware.handle(this::handleKeyRefresh), Role.OPERATOR));
         router.get("/key/acl/refresh").handler(auth.handle(attestationMiddleware.handle(this::handleKeyAclRefresh), Role.OPERATOR));
+        router.get("/key/keyset/refresh").handler(auth.handle(attestationMiddleware.handle(this::handleKeysetRefresh), Role.OPERATOR));
+        router.get("/key/keyset_keys/refresh").handler(auth.handle(attestationMiddleware.handle(this::handleKeysetKeyRefresh), Role.OPERATOR));
```
Comment (aulme) [inline src/main/java/com/uid2/core/vertx/CoreVerticle.java:134]: Are we settled on the URL? Generally dashes are the idiomatic way to separate words in a URL

File: src/main/java/com/uid2/core/service/KeysetKeysMetadataProvider.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import com.uid2.shared.cloud.ICloudStorage;
+import com.uid2.core.model.SecretStore;
+import io.vertx.core.json.Json;
+import io.vertx.core.json.JsonObject;
+
+
+import static com.uid2.core.util.MetadataHelper.getMetadataPathName;
+import static com.uid2.core.util.MetadataHelper.readToEndAsString;
+
+public class KeysetKeysMetadataProvider implements IKeysetKeyMetadataProvider {
```
Comment (aulme) [inline src/main/java/com/uid2/core/service/KeysetKeysMetadataProvider.java:14]: np: the interface name looks like it's intended to match the class name but doesn't

Also I feel the name is too easy to confuse with the KeysetMetadataProvider - but can't think of a  better one so unless someone has a better idea may be ok 🤔

File: src/main/java/com/uid2/core/service/KeysetKeysMetadataProvider.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    private final ICloudStorage downloadUrlGenerator;
+
+    public KeysetKeysMetadataProvider(ICloudStorage cloudStorage) {
+        this.metadataStreamProvider = this.downloadUrlGenerator = cloudStorage;
+    }
+
+    @Override
+    public String getMetadata(OperatorInfo info) throws Exception {
+        String pathname = getMetadataPathName(info.getOperatorType(), info.getSiteId(), SecretStore.Global.get(Const.Config.KeysetKeysMetadataPathProp));
+        String original = readToEndAsString(metadataStreamProvider.download(pathname));
```
Comment (aulme) [inline src/main/java/com/uid2/core/service/KeysetKeysMetadataProvider.java:25]: I feel the names "main" and "obj" are not too useful. Perhaps "original" could be renamed to "originalRaw", "main" to "original", and "obj" to "keysetKeys"?

File: src/main/java/com/uid2/core/service/KeysetKeysMetadataProvider.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+public class KeysetKeysMetadataProvider implements IKeysetKeyMetadataProvider {
+    private final ICloudStorage metadataStreamProvider;
+    private final ICloudStorage downloadUrlGenerator;
+
+    public KeysetKeysMetadataProvider(ICloudStorage cloudStorage) {
+        this.metadataStreamProvider = this.downloadUrlGenerator = cloudStorage;
+    }
+
+    @Override
+    public String getMetadata(OperatorInfo info) throws Exception {
```
Comment (aulme) [inline src/main/java/com/uid2/core/service/KeysetKeysMetadataProvider.java:23]: Looks like this `getMetadata` is the same for data types, just with a different metadata path and property name. Could we extract this into a reused helper?

File: src/main/java/com/uid2/core/service/KeysetKeysMetadataProvider.java (hunk 4 above)
Comment (jon8787) [reply src/main/java/com/uid2/core/service/KeysetKeysMetadataProvider.java:14]: Do we need IKeysetKeyMetadataProvider at all? 🤔

File: src/main/java/com/uid2/core/util/MetadataHelper.java [hunk 7]
```diff
@@ -33,4 +36,14 @@ public static String getMetadataPathName(OperatorType operatorType, int siteId,
         }
         return store.getMetadataPath().toString();
     }
+
+    public static String readToEndAsString(InputStream stream) throws IOException {
+        final InputStreamReader reader = new InputStreamReader(stream);
+        final char[] buff = new char[1024];
```
[the remark this answers is not in the document]
Comment (cody-constine-ttd) [reply src/main/java/com/uid2/core/util/MetadataHelper.java:42]: It is old code and its just reading the metadata file which appear to max out at 125 chars and don't grow much from that. We could restructure this to use an array?

**Review by jon8787 — APPROVED** [review]

**Pull request merged**